### PR TITLE
Pin actions to a full length commit SHA

### PR DESCRIPTION
**Pull request by naveensrinivasan**
- Pinned actions by SHA https://github.com/ossf/scorecard/blob/main/docs/checks.md#pinned-dependencies
- Included permissions for the action. https://github.com/ossf/scorecard/blob/main/docs/checks.md#token-permissions

>Pin actions to a full length commit SHA

>Pinning an action to a full length commit SHA is currently the only way to use an action as an immutable release. Pinning to a particular SHA helps mitigate the risk of a bad actor adding a backdoor to the action's repository, as they would need to generate a SHA-1 collision for a valid Git object payload.

https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-third-party-actions

Also, dependabot supports upgrading based on SHA.

- Also included permissions for actions to reduce the attack vector.

GitHub's own repository pin's their `checkout` actions by SHA and doesn't use the version tag https://github.com/github/docs/blob/ea7f218c91ecbae9a700a8702b51a7d2736e0d2c/.github/workflows/docs-review-collect.yml#L23

Signed-off-by: naveensrinivasan <172697+naveensrinivasan@users.noreply.github.com>

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by naveensrinivasan** [comment]
@htuch 👀 please.

**Comment by htuch** [comment]
Looks good in principle, over to @phlax to double check and merge. I'm not sure I fully grok every one of the permissions.

**Comment by phlax** [comment]
as a general comment - the downside to using shas is that they are not human readable - but i would agree that this is preferable from a security pov - we have done this already with some of the actions etc

adding the version as a comment after is not so helpful as dependabot doesnt update this so it goes stale very quickly

ill look over the permissions

**Comment by phlax** [comment]
we might also want to add this https://github.com/marketplace/actions/ensure-sha-pinned-actions to ensure that sha-pinning is enforced moving forward

File: .github/workflows/check-deps.yml [hunk 1]
```diff
@@ -16,7 +16,7 @@ jobs:
         with:
           ref: ${{ github.head_ref }}
       - name: Set up Python ${{ matrix.python-version }}
-        uses: actions/setup-python@v3
+        uses: actions/setup-python@0ebf233433c08fb9061af664d501c3f3ff0e9e20 # v3
```
Comment (phlax) [inline .github/workflows/check-deps.yml:19]: i think remove the v3 from the end as it will not get updated

File: .github/workflows/check-deps.yml [hunk 2]
```diff
@@ -16,7 +16,7 @@ jobs:
         with:
```
Comment (phlax) [inline .github/workflows/check-deps.yml:16]: possibly i think we also want to add permissions for this workflow

File: .github/workflows/codeql-daily.yml [hunk 3]
```diff
@@ -2,9 +2,16 @@ on:
   schedule:
   - cron: '0 12 * * 4'
 
+permissions:
+  contents: read
```
Comment (phlax) [inline .github/workflows/codeql-daily.yml:6]: im not sure that this is necessary

File: .github/workflows/codeql-daily.yml [hunk 4]
```diff
@@ -2,9 +2,16 @@ on:
...
+
 jobs:
   CodeQL-Build:
 
+    permissions:
+      actions: read  # for github/codeql-action/init to get workflow details
+      contents: read  # for actions/checkout to fetch code
```
Comment (phlax) [inline .github/workflows/codeql-daily.yml:13]: i dont think these 2 are necessary - checking the codeql readme it suggests these are only required for private repos

File: .github/workflows/codeql-push.yml [hunk 5]
```diff
@@ -8,9 +8,16 @@ on:
     - 'dependabot/**'
   pull_request:
 
+permissions:
+  contents: read
```
Comment (phlax) [inline .github/workflows/codeql-push.yml:12]: i think this is unnecessary

Comment (phlax) [reply .github/workflows/codeql-push.yml:12]: same elsewhere

File: .github/workflows/codeql-push.yml [hunk 6]
```diff
@@ -8,9 +8,16 @@ on:
...
+
 jobs:
   CodeQL-Build:
 
+    permissions:
+      actions: read  # for github/codeql-action/init to get workflow details
+      contents: read  # for actions/checkout to fetch code
```
Comment (phlax) [inline .github/workflows/codeql-push.yml:19]: as above re these being unnecessary

File: .github/workflows/pr_notifier.yml [hunk 7]
```diff
@@ -12,7 +15,7 @@ jobs:
     steps:
     - uses: actions/checkout@a12a3943b4bdde767164f792f33f40b04645d846
     - name: Set up Python 3.8
-      uses: actions/setup-python@v3
+      uses: actions/setup-python@0ebf233433c08fb9061af664d501c3f3ff0e9e20 # v3
```
Comment (phlax) [inline .github/workflows/pr_notifier.yml:18]: remove v3 suffix

File: .github/workflows/pr_notifier.yml [hunk 8]
```diff
@@ -3,6 +3,9 @@ on:
   schedule:
   - cron: '0 5 * * 1,2,3,4,5'
 
+permissions:
+  contents: read
```
Comment (phlax) [inline .github/workflows/pr_notifier.yml:7]: in this case i think we want to remove all permissions - ie `permissions: {}`

File: .github/workflows/check-deps.yml (hunk 2 above)
Comment (phlax) [reply .github/workflows/check-deps.yml:16]: atm i think this one just needs `issues: write`

**Review by phlax — CHANGES_REQUESTED** [review]
@naveensrinivasan thanks for this - its a welcome addition

before landing it would be good to tighten these so we only add the required permissioning and conversely tighten any that dont need (any) permissioning

File: .github/workflows/check-deps.yml (hunk 1 above)
Comment (naveensrinivasan) [reply .github/workflows/check-deps.yml]: Done.

File: .github/workflows/check-deps.yml (hunk 2 above)
Comment (naveensrinivasan) [reply .github/workflows/check-deps.yml:16]: Done.

File: .github/workflows/codeql-daily.yml (hunk 3 above)
Comment (naveensrinivasan) [reply .github/workflows/codeql-daily.yml:6]: Removed. Thanks

File: .github/workflows/codeql-daily.yml (hunk 4 above)
Comment (naveensrinivasan) [reply .github/workflows/codeql-daily.yml]: Done.

File: .github/workflows/pr_notifier.yml (hunk 7 above)
Comment (naveensrinivasan) [reply .github/workflows/pr_notifier.yml]: Done.

File: .github/workflows/pr_notifier.yml (hunk 8 above)
Comment (naveensrinivasan) [reply .github/workflows/pr_notifier.yml]: Removed.

File: .github/workflows/codeql-push.yml (hunk 6 above)
Comment (naveensrinivasan) [reply .github/workflows/codeql-push.yml]: Done.

**Comment by naveensrinivasan** [comment]
> @naveensrinivasan thanks for this - its a welcome addition
> 
> before landing it would be good to tighten these so we only add the required permissioning and conversely tighten any that dont need (any) permissioning

I have taken care of everything except adding additional permissions that were recommended. I am not sure that is the set of permissions that are required. I would suggest being a separate PR after this one is merged. 

Thanks for the review and comments.

**Comment by phlax** [comment]
>  I am not sure that is the set of permissions that are required.

I am reasonably confident that is what is required, but as it happens by schedule rather than trigger im happy to address separately

> I would suggest being a separate PR after this one is merged.

Cool. do you want to pick that up - im happy to do it if not (just dont want to forget about it)

File: .github/workflows/pr_notifier.yml [hunk 9]
```diff
@@ -12,7 +13,7 @@ jobs:
     steps:
     - uses: actions/checkout@a12a3943b4bdde767164f792f33f40b04645d846
     - name: Set up Python 3.8
-      uses: actions/setup-python@v3
+      uses: actions/setup-python@0ebf233433c08fb9061af664d501c3f3ff0e9e20
```
Comment (phlax) [inline .github/workflows/pr_notifier.yml:16]: i think this will fail formatting due to trailing ws

File: .github/workflows/stale.yml [hunk 10]
```diff
@@ -3,15 +3,21 @@ on:
   schedule:
   - cron: '0 */4 * * *'
 
+permissions:
```
Comment (phlax) [inline .github/workflows/stale.yml:6]: lets remove this one also - i dont think we need any read permissions as its a public repo

File: .github/workflows/codeql-daily.yml (hunk 3 above)
Comment (phlax) [reply .github/workflows/codeql-daily.yml:6]: seems still here

File: .github/workflows/pr_notifier.yml (hunk 9 above)
Comment (phlax) [reply .github/workflows/pr_notifier.yml:16]: copying out ci advice re editor/formatting

```console
 Please fix your editor to ensure:

      - no trailing whitespace
      - no preceding mixed tabs/spaces
      - all files end with a newline

```

File: .github/workflows/pr_notifier.yml [hunk 11]
```diff
@@ -3,6 +3,7 @@ on:
   schedule:
   - cron: '0 5 * * 1,2,3,4,5'
 
+
```
Comment (phlax) [inline .github/workflows/pr_notifier.yml:6]: nit: unnecessary change/ws - ditto elsewhere

File: .github/workflows/pr_notifier.yml (hunk 9 above)
Comment (naveensrinivasan) [reply .github/workflows/pr_notifier.yml]: Took care of it. Thanks

File: .github/workflows/stale.yml (hunk 10 above)
Comment (naveensrinivasan) [reply .github/workflows/stale.yml]: Got it. Thanks

**Comment by naveensrinivasan** [comment]
> > I am not sure that is the set of permissions that are required.
> 
> I am reasonably confident that is what is required, but as it happens by schedule rather than trigger im happy to address separately
> 
> > I would suggest being a separate PR after this one is merged.
> 
> Cool. do you want to pick that up - im happy to do it if not (just dont want to forget about it)

Thanks, I would rather have you do it :)

File: .github/workflows/codeql-daily.yml (hunk 3 above)
Comment (naveensrinivasan) [reply .github/workflows/codeql-daily.yml]: Got it.

**Review by phlax — APPROVED** [review]
great. lgtm thanks @naveensrinivasan for iterating

**Comment by naveensrinivasan** [comment]
> great. lgtm thanks @naveensrinivasan for iterating

👍
